### PR TITLE
[pytree] align function signature between C++ and Python pytree

### PR DESCRIPTION
**Pull request by XuehaiPan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112477
* #112485
* #112278
* #112111
* __->__ #112482

Change the argument name in C++ and Python pytree APIs. Also add a test to ensure the function signatures are the same in the two implementations.

- #112485

cc @zou3519 @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

### CODE REVIEW

**Comment by zou3519** [comment]
> Change the argument name in C++ and Python pytree APIs. Also add a test to ensure the function signatures are the same in the two implementations.

Where is the test?

File: torch/utils/_cxx_pytree.py [hunk 1]
```diff
@@ -475,6 +476,11 @@ def map_only(__type_or_types: Type2[T, S]) -> MapOnlyFn[Fn2[T, S, Any]]:
     ...
 
 
+@overload
+def map_only(__type_or_types: Type3[T, S, U]) -> MapOnlyFn[Fn3[T, S, U, Any]]:
```
Comment (zou3519) [inline torch/utils/_cxx_pytree.py:480]: For my understanding: what is `Type3[T, S, U]`?

File: torch/utils/_pytree.py [hunk 2]
```diff
@@ -481,99 +487,168 @@ def go(t):
...
     """
 
-    def deco(f: Callable[[T], Any]) -> Callable[[Any], Any]:
-        def inner(x: T) -> Any:
-            if isinstance(x, ty):
-                return f(x)
-            else:
-                return x
+    def wrapper(func: Callable[[T], Any]) -> Callable[[Any], Any]:
+        # @functools.wraps(func)  # torch dynamo doesn't support this yet
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/utils/_pytree.py:491]: We shouldn't in this PR. Open question on if we should do it in the future.

**Review by zou3519 — COMMENTED** [review]
we probably should not rename the args of `_register_pytree_node`. Other than that, this looks fine. I assume a test is coming later?

**Comment by XuehaiPan** [comment]
@zou3519

> we probably should not rename the args of `_register_pytree_node`.

If we prefer `fn` over `func` in argument names, I can change all these to `fn`.

> Other than that, this looks fine. I assume a test is coming later?

The tests are added in:

- #112485

File: torch/utils/_cxx_pytree.py (hunk 1 above)
Comment (XuehaiPan) [reply torch/utils/_cxx_pytree.py:480]: It's a 3-tuple of types. For example:

```python
isinstace(x, (int, float, torch.Tensor))
             ^------------------------- Type3[int, float, torch.Tensor]
```

**Comment by zou3519** [comment]
> > we probably should not rename the args of `_register_pytree_node`.
> 
> If we prefer `fn` over `func` in argument names, I can change all these to `fn`.

I'm trying to not break backwards compatibility. The args of `_register_pytree_node` is the one that I'm most nervous about, everything else seems OK.

IMO we should leave `_register_pytree_node` alone right now. It's a private API anyways (prefixed with the underscore). In the future we'll have some public `register_pytree_node` that we have whatever signature we deem good.

**Comment by XuehaiPan** [comment]
> I'm trying to not break backwards compatibility. The args of `_register_pytree_node` is the one that I'm most nervous about, everything else seems OK.
> 
> IMO we should leave `_register_pytree_node` alone right now. It's a private API anyways (prefixed with the underscore). In the future we'll have some public `register_pytree_node` that we have whatever signature we deem good.

I changed the argument name to `*_fn` back in `register_pytree_node`. I leave the change to `func` in `tree_map` because it is always called via positional argument.

IMO, we should use a unified pattern for these argument names. In Python pytree and fx pytree we use `fn` in argument names but use `Func` in type hints.

https://github.com/pytorch/pytorch/blob/a1ab22b81d74a26873243d766d78724b3b70d570/torch/utils/_pytree.py#L125-L126

https://github.com/pytorch/pytorch/blob/a1ab22b81d74a26873243d766d78724b3b70d570/torch/fx/_pytree.py#L16

https://github.com/pytorch/pytorch/blob/85e93632e7804bfe64316cbc491aa803a68b0701/torch/_export/utils.py#L25-L26

https://github.com/pytorch/pytorch/blob/85e93632e7804bfe64316cbc491aa803a68b0701/torch/onnx/_internal/fx/dynamo_graph_extractor.py#L53-L55

**Review by zou3519 — APPROVED** [review]

**Comment by zou3519** [comment]
> IMO, we should use a unified pattern for these argument names. In Python pytree and fx pytree we use fn in argument names but use Func in type hints.

I agree in general, though we have to be careful of BC at times.

**Comment by XuehaiPan** [comment]
Hi @PaliC, the only signature change is renaming the first argument from `typ` to `cls`. This argument is always passed as positional argument.

**Pull request closed without merging**